qcoro/impl/waitfor.h

```diff
@@ -73,15 +86,27 @@ T waitFor(Awaitable &&awaitable) {
             std::rethrow_exception(context.exception);
         }
     } else {
-        T result;
-        runCoroutine(context, result, std::forward<Awaitable>(awaitable));
-        if (!context.coroutineFinished) {
-            context.loop.exec();
-        }
-        if (context.exception) {
-            std::rethrow_exception(context.exception);
+        if constexpr (std::is_default_constructible_v<T>) {
+            T result;
+            runCoroutine(context, result, std::forward<Awaitable>(awaitable));
+            if (!context.coroutineFinished) {
+                context.loop.exec();
+            }
+            if (context.exception) {
+                std::rethrow_exception(context.exception);
+            }
+            return result;
+        } else {
+            std::optional<T> result;
+            runCoroutineOpt(context, result, std::forward<Awaitable>(awaitable));
+            if (!context.coroutineFinished) {
+                context.loop.exec();
+            }
+            if (context.exception) {
+                std::rethrow_exception(context.exception);
+            }
+            return *result;
```

Comment:
You should be able to `std::move` the value here. You are right that we lose RVO with the optional, but that should be a negligible cost, especially if we consider the overall overhead of waitFor due to running a nested QEventLoop and suspending/resuming coroutines themselves.

Reply:
added explicit move to hopefully be more effienct